campaignion_tracking/js/campaignion_tracking.js

```diff
@@ -0,0 +1,123 @@
+(function($) {
+Drupal.behaviors.campaignion_tracking = {};
+Drupal.behaviors.campaignion_tracking.attach = function(context, settings) {
+
+  /**
+   * Dispatch to campaignion_tracking.tracker.
+   *
+   * Nothing happens if the tracker is not available.
+   */
+
+  function gracefulDispatch (name, data, trackingContext) {
+    // Build an msg object expected by campaignion_tracking:
+    // It has a "name" for the event, "data" for the event and a
+    // "context" for additional data.
+    // NB: the event handler of the tracker might expect a certain data
+    // structure.
+    var msg = {
+      'name': name,
+      'data': data,
+      'context': trackingContext
+    }
+    if (typeof window.campaignion_tracking.tracker !== 'undefined') {
+      window.campaignion_tracking.tracker.publish('donation', msg)
+    }
+  }
+
+  /**
+   * Donation related events to trigger on page load.
+   *
+   * The events are only fired if a correct campaignion_tracking context is
+   * available. They are only fired once for the Drupal context to prevent the
+   * same messages being dispatch multiple times (behaviours can be called
+   * multiple times).
+   *
+   * Events:
+   * - 'donationSuccess'
+   *
+   * TODO: Missing events for impression and detail view
+   */
+  $('html', context).once('campaignion-tracking', function () {
+    if (settings['campaignion_tracking'] &&
+        settings['campaignion_tracking']['context'] &&
+        settings['campaignion_tracking']['context']['submitted']) {
+      var node = settings.campaignion_tracking.context.submitted['node'] || {};
+      var donation = settings.campaignion_tracking.context.submitted['donation'] || {};
+      if (node['is_donation'] &&
+          donation['amount'] &&
+          donation['interval']) {
+        var product = {
+          name: node['title'],
+          id: node['nid'],
+          price: String(donation['amount']),
+          variant: String(donation['interval']),
+          quantity: 1
+        };
+        // TODO change name, place
+        var msg = {
+          currencyCode: donation['currency_code'] || 'EUR',
+          product: product,
+          tid: donation['pid']
+        };
+        gracefulDispatch('donationSuccess', msg, settings.campaignion_tracking.context);
+      }
+    }
+  });
+
+  /**
+   * Donation related events to trigger on webform submits.
+   *
+   * The events are only fired if a correct campaignion_tracking context is
+   * available. They are only fired once for the Drupal context to prevent the
+   * same messages being dispatch multiple times (behaviours can be called
+   * multiple times).
+   *
+   * Events:
+   * - 'setDonationProduct'
+   * - 'checkoutBegin'
+   * - 'checkoutEnd'
+   */
+  $('.webform-client-form', context).once('campaignion-tracking', function () {
```

Comment:
The once hinders any event from being dispatched that is not in the `settings` the first time the behaviour is called. I don’t think this is the intended behaviour.

Reply:
this is called once per page load (when not using webform AJAX) or once per AJAX form submit.
AJAX form submits completely replace the `<form>` markup, thus this handler get's called exactly once per (successful) submit on multistep forms.

The reason for this once is that otherwise the behaviour get's called *multiple* times *per* 1 submit.

Maybe there is a better way. But it does what it is supposed to to (as far as I have tested)...

Reply:
reworked. `once()` is gone now. introduced `settings['campaignion_tracking']['sent']` to be sure we are not sending the same events twice...